merkle/coniks/coniks.go

```diff
@@ -0,0 +1,124 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package coniks provides hashing for maps.
+package coniks
+
+import (
+	"crypto"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/merkle/hashers"
+)
+
+func init() {
+	hashers.RegisterMapHasher(trillian.HashStrategy_CONIKS_SHA512_256, Default)
+}
+
+// Domain separation prefixes
+var (
+	leafIdentifier  = []byte("L")
+	emptyIdentifier = []byte("E")
+)
+
+// Default is the standard CONIKS hasher.
+var Default = New(crypto.SHA512_256)
+
+// hasher implements the sparse merkle tree hashing algorithm specified in the CONIKS paper.
+type hasher struct {
+	crypto.Hash
+}
+
+// New creates a new hashers.TreeHasher using the passed in hash function.
+func New(h crypto.Hash) hashers.MapHasher {
+	return &hasher{Hash: h}
+}
+
+// EmptyRoot returns the root of an empty tree.
+func (m *hasher) EmptyRoot() []byte {
+	panic("EmptyRoot() not defined for coniks.Hasher")
+}
+
+// HashEmpty returns the hash of an empty branch at a given height.
+// A height of 0 indicates the hash of an empty leaf.
+// Empty branches within the tree are plain interior nodes e1 = H(e0, e0) etc.
+func (m *hasher) HashEmpty(treeID int64, index []byte, height int) []byte {
+	depth := m.BitLen() - height
+
+	h := m.New()
+	h.Write(emptyIdentifier)
+	binary.Write(h, binary.BigEndian, uint64(treeID))
+	h.Write(m.maskIndex(index, depth))
+	binary.Write(h, binary.BigEndian, uint32(depth))
+	return h.Sum(nil)
+}
+
+// HashLeaf calculate the merkle tree leaf value:
+// H(Identifier || treeID || depth || index || dataHash)
+func (m *hasher) HashLeaf(treeID int64, index []byte, height int, leaf []byte) []byte {
+	depth := m.BitLen() - height
+
+	h := m.New()
+	h.Write(leafIdentifier)
+	binary.Write(h, binary.BigEndian, uint64(treeID))
+	h.Write(m.maskIndex(index, depth))
+	binary.Write(h, binary.BigEndian, uint32(depth))
+	h.Write(leaf)
+	return h.Sum(nil)
+}
+
+// HashChildren returns the internal Merkle tree node hash of the the two child nodes l and r.
+// The hashed structure is  H(l || r).
+func (m *hasher) HashChildren(l, r []byte) []byte {
+	h := m.New()
+	h.Write(l)
+	h.Write(r)
+	return h.Sum(nil)
+}
+
+// BitLen returns the number of bits in the hash function.
+func (m *hasher) BitLen() int {
+	return m.Size() * 8
+}
+
+// leftmask contains bitmasks indexed such that the left x bits are set.
+// 0 is special cased to 0xFF since 8 mod 8 is 0.
+var leftmask = map[int]byte{
+	0: 0xFF, 1: 0x80, 2: 0xC0, 3: 0xE0,
+	4: 0xF0, 5: 0xF8, 6: 0xFC, 7: 0xFE,
+}
+
+// maskIndex returns index with only the left depth bits set.
+// index must be of size m.Size() and 0 <= depth <= m.BitLen().
```

Comment:
This could probably be an array. I think including an example in the comment would help here. Maybe also mention that it's only used for the last byte to explain the 0xff in position 0.